### PR TITLE
github: try out golangci-lint

### PR DESCRIPTION
**Pull request by bboozzoo**
Try out golangci-lint linter.

### CODE REVIEW

File: cmd/snap/main.go [hunk 1]
```diff
@@ -591,3 +591,7 @@ fixed-width fonts, so it can be hard to tell.
 
 	return nil
 }
```
Comment (bboozzoo) [inline cmd/snap/main.go:593]: This bit is only here to demonstrate automatic annotations.

File: .golangci.yml [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+  disable-all: true
+
+
+issues:
+  # List of regexps of issue texts to exclude, empty list by default.
+  # But independently from this option we use default exclude patterns,
+  # it can be disabled by `exclude-use-default: false`. To list all
+  # excluded by default patterns execute `golangci-lint run --help`
+  exclude:
+    - abcdef
```
Comment (anonymouse64) [inline .golangci.yml:198]: what does this do?

**Review by anonymouse64 — COMMENTED** [review]
lgtm, I really like the annotations here, but a couple questions about the settings for the golangci.yml defaults

File: .golangci.yml [hunk 3]
```diff
@@ -0,0 +1,236 @@
+# This file contains all available configuration options
+# with their default values.
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply .golangci.yml:2]: This based on the config I've been using for a longer while now.

[the remark this answers is not in the document]
Comment (bboozzoo) [reply .golangci.yml:2]: FWIW a full example config is right here: https://github.com/golangci/golangci-lint/blob/master/.golangci.example.yml

File: .golangci.yml [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+  # which files to skip: they will be analyzed, but issues from them
+  # won't be reported. Default value is empty list, but there is
+  # no need to include all autogenerated files, we confidently recognize
+  # autogenerated files. If it's not please let us know.
+  # "/" will be replaced by current OS file path separator to properly work
+  # on Windows.
+  # skip-files:
+  #   - export_test.go
+
+  # by default isn't set. If set we pass it to "go list -mod={option}". From "go help modules":
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply .golangci.yml:45]: Actually, it's like this in the upstream config too.

File: .golangci.yml [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+      - auther
+      - PROCES
+      - PROCESSS
+      - proces
+      - processs
+      - exportfs
+  lll:
+    # max line length, lines longer will be reported. Default is 120.
+    # '\t' is counted as 1 character by default, and can be changed with the tab-width option
+    line-length: 120
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply .golangci.yml:149]: I'm using 80 too, but I didn't want to enforce anything too controversial here.

File: .golangci.yml [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    - misspell
+    - deadcode
+    - unused
+    - gosimple
+    - nakedret
+    - gofmt
+    - ineffassign
+    - goimports
+    - testpackage
+  disable-all: true
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply .golangci.yml:189]: Yes, we disable all the linters and then enable only select few. I've added a comment.

File: .golangci.yml (hunk 2 above)
Comment (bboozzoo) [reply .golangci.yml:198]: Ah, forgot to comment it out. This allows us to selectively skip some issues by listing matching regular expressions.

**Review by mvo5 — APPROVED** [review]
I like this! Let's remove the cmd/snap/main.go change and then it got my :+1:

File: .golangci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (mardy) [reply .golangci.yml:145]: +1 for 80 :-)

File: .golangci.yml [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    - govet
+    - misspell
+    - deadcode
+    - unused
+    - gosimple
+    - nakedret
+    - gofmt
+    - ineffassign
+    - goimports
+    - testpackage
```
Comment (mardy) [inline .golangci.yml:185]: I think that the goal of this linter is valuable, but our current code base is very far from satisfying it: we are using whitebox testing (testing of internal functions) pretty much everywhere.

According to the documentation of the `testpackage` linter, our issues could be solved if we moved the whitebox tests into files called `XXX_internal_test.go`. Do we want to do it?

Comment (mardy) [reply .golangci.yml:185]: Actually, I missed Ian's [comment to a PR of mine](https://github.com/snapcore/snapd/pull/10234#issuecomment-832179286), where he suggests how these errors should be fixed. I'll give that a try :-)

File: .golangci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply .golangci.yml:145]: well, with their default tab-width, 80 doesn't mean really 80, anyway these days even black doesn't default to 80. I need to think about this.

File: .golangci.yml [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    for-loops: false # Report preallocation suggestions on for loops, false by default
+
+linters:
+  # enable the following linters
+  enable:
+    - govet
+    - misspell
+    - deadcode
+    - unused
+    - gosimple
```
Comment (pedronis) [inline .golangci.yml:180]: I find gosimple suggestions both potentially useful but also a bit annoying

File: .golangci.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply .golangci.yml:145]: this linter is not part of the list below so it doesn't matter afaict

File: .github/workflows/test.yaml [hunk 9]
```diff
@@ -138,6 +142,26 @@ jobs:
...
     - name: Get Go dependencies
       run: cd ${{ github.workspace }}/src/github.com/snapcore/snapd && ${{ github.workspace }}/bin/govendor sync
+
+    - name: golangci-lint
+      uses: golangci/golangci-lint-action@v2
+      if: ${{ matrix.gochannel == 'latest/stable' }}
+      with:
+        # version of golangci-lint to use in form of v1.2 or v1.2.3 or `latest`
+        # to use the latest version
+        version: latest
```
Comment (pedronis) [inline .github/workflows/test.yaml:152]: we probably shouldn't do this and pick a version instead and upgrade once in a while. We don't want to get annoyed/blocked by linter changes

**Review by pedronis — COMMENTED** [review]
my input on this so far

File: .golangci.yml (hunk 8 above)
Comment (pedronis) [reply .golangci.yml:180]: Also I don't think we can take some of the recommendations until we have switched to go1.13? some seems to be about formatting

File: .golangci.yml (hunk 7 above)
Comment (bboozzoo) [reply .golangci.yml:185]: Tweaked the file patterns used when excluding testpackage.

**Comment by pedronis** [comment]
@bboozzoo 

it's a bit weird that the  output doesn't seem to be sorted, if you run the thing many times you get same output(?) but in different order? oh well

googling very quickly it seems goimports also does gofmt checks of some kind as well??

we get errors like:

```
overlord/servicestate/service_control_test.go:863: File is not `goimports`-ed with -local github.com/snapcore/snapd (goimports)
			enable: []string{"a"},
```

it seems we need to exclude osutil/udev also from nakedret, mayebe we should exclude completely


You said this will analyze only changed files?

We still have too many test files testpackage is unhappy about, might we should list them for exclusion for it explicitly?

**Comment by pedronis** [comment]
so I checked goimports expects the gofmt outcome, I suppose same as go version it was compiled with, this means we probably would need to switch locally to use goimports compiled with the same of go as the golangci-lint we use instead of gofmt?

**Comment by bboozzoo** [comment]
> @bboozzoo
> 
> it's a bit weird that the output doesn't seem to be sorted, if you run the thing many times you get same output(?) but in different order? oh well
> 

Yeah, I suppose the results aren't really sorted by the kind of the linter.

> 
> we get errors like:
> 
> ```
> overlord/servicestate/service_control_test.go:863: File is not `goimports`-ed with -local github.com/snapcore/snapd (goimports)
> 			enable: []string{"a"},
> ```
> 
> it seems we need to exclude osutil/udev also from nakedret.
> 
> You said this will analyze only changed files?

Yes, but this applies only to the changes on github and uses master as reference. When you run locally, you can pass `--new-from-rev=<commit-sha>`.

> 
> We still have too many test files testpackage is unhappy about, might we should list them for exclusion for it explicitly?

I'll look into this. It should be possible to whiltelist those files.

**Review by pedronis — APPROVED** [review]
this should be usable for us now, we will need to consider fixing slowly the testpackage issues and move back to goimports once we use a more modern  baseline formatting.

@bboozzoo as discussed please double check what I did

**Comment by bboozzoo** [comment]
:+1: for the changes from @pedronis

**Pull request merged**